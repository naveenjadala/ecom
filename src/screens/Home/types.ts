export interface ArtObjects {
  id: string;
  objectNumber: string;
  title: string;
}
export interface ArtData {
  artObjects: ArtObjects[];
}
