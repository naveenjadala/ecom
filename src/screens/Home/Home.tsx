import {ActivityIndicator, Button, Text, View} from 'react-native';
import React, {useEffect, useState} from 'react';
import {useNavigation} from '@react-navigation/native';
import Icon from 'react-native-vector-icons/FontAwesome';
import isEqual from 'lodash/isEqual';

import {useApiCall} from '../../networking/useApiCall';
import {getArtgallery} from '../../networking/urls';
import CustomFlatList from '../../components/customLists/CustomFlatList';
import {ArtData, ArtObjects} from './types';
import {styles} from './styles';
import RenderItems from './components/RenderItems';
import HomeIcon from '../../assets/images/home.svg';
import {HomeStackNavigationProp} from '../../navigation/Route';
import {TouchableOpacity} from 'react-native';

/**
 * Renders the Home component.
 *
 * @returns {JSX.Element} The rendered Home component.
 */
const Home = () => {
  const {loading, error, data, getApi} = useApiCall<ArtData>();
  const [artData, setArtData] = useState<ArtObjects[]>([]);
  const [page, setPage] = useState(1);
  const {navigate} = useNavigation<HomeStackNavigationProp>();

  useEffect(() => {
    getApi(getArtgallery.getAllArts, {key: 'GMZbiDud', p: page});
  }, [page]);

  useEffect(() => {
    if (data?.artObjects && !isEqual(data?.artObjects, artData)) {
      setArtData(prevData => [...prevData, ...data?.artObjects]);
    }
  }, [data, artData]);

  /**
   * Loads more data if scroll to the bottom.
   *
   * @return {void}
   */
  const loadMoreData = () => {
    if (!loading) {
      setPage(prevData => prevData + 1);
    }
  };

  // Navigates to the 'Details' screen.
  const navigateToDetails = () => {
    navigate('Details');
  };

  return (
    <View style={{...styles.container}}>
      <Icon name="rocket" size={30} color="#900" />
      {/* <Button testID="button" title="click" onPress={navigateToDetails} /> */}
      <TouchableOpacity accessibilityRole="button" onPress={navigateToDetails}>
        <Text>press</Text>
      </TouchableOpacity>
      <HomeIcon height={20} width={20} />
      <CustomFlatList
        data={artData}
        renderItem={RenderItems}
        keyExtractor={item => item.id}
        ListEmptyComponent={() =>
          (!loading && <Text>No data found</Text>) || null
        }
        loadMoreData={loadMoreData}
        ListFooterComponent={() =>
          (loading && <ActivityIndicator size={30} />) || null
        }
      />
    </View>
  );
};

export default Home;
