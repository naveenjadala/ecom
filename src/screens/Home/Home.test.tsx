import {useNavigation} from '@react-navigation/native';
import {
  fireEvent,
  render,
  screen,
  useNavigationMock,
  userEvent,
  waitFor,
} from '../../test/test-utils';
import Home from './Home';

jest.mock('@react-navigation/native', () => ({
  ...jest.requireActual('@react-navigation/native'),
  useNavigation: jest.fn(() => ({navigate: jest.fn()})), // Ensure useNavigation returns an object with a navigate property
}));

// Mock the navigation hook
// jest.mock('@react-navigation/native', () => ({
//   ...jest.requireActual('@react-navigation/native'),
//   useNavigation: jest.fn(),
// }));

// Mock the useApiCall hook
jest.mock('../../networking/useApiCall', () => ({
  useApiCall: jest.fn(() => ({
    loading: false,
    error: null,
    data: {
      artObjects: [
        {id: '1', title: 'Art 1'},
        {id: '2', title: 'Art 2'},
      ],
    },
    getApi: jest.fn(),
  })),
}));

describe('Home', () => {
  test('should render Home screen', () => {
    render(<Home />);
    expect(screen).toMatchSnapshot();
  });

  test('should render Home screen with data from API', async () => {
    render(<Home />);
    expect(screen.getByText('Art 1')).toBeTruthy();
    expect(screen.getByText('Art 2')).toBeTruthy();
  });

  test('should render Home screen with no data from API', async () => {
    (require('../../networking/useApiCall') as any).useApiCall.mockReturnValue({
      loading: false,
      error: null,
      data: {artObjects: []},
      getApi: jest.fn(),
    });
    render(<Home />);
    expect(screen.getByText('No data found')).toBeOnTheScreen();
  });
});
