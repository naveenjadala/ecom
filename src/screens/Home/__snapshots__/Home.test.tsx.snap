// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should render Home screen 1`] = `
<RCTSafeAreaView
  style={
    {
      "backgroundColor": "#fff",
      "flex": 1,
      "minHeight": "100%",
      "minWidth": "100%",
    }
  }
>
  <View
    style={
      {
        "flex": 1,
      }
    }
  >
    <Text
      allowFontScaling={false}
      selectable={false}
      style={
        [
          {
            "color": "#900",
            "fontSize": 30,
          },
          undefined,
          {
            "fontFamily": "FontAwesome",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          {},
        ]
      }
    >
      ï„µ
    </Text>
    <View
      accessibilityRole="button"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <Text>
        press
      </Text>
    </View>
    <test-file-stub
      height={20}
      width={20}
    />
    <RCTScrollView
      ListEmptyComponent={[Function]}
      ListFooterComponent={[Function]}
      data={
        [
          {
            "id": "1",
            "title": "Art 1",
          },
          {
            "id": "2",
            "title": "Art 2",
          },
        ]
      }
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onEndReached={[Function]}
      onEndReachedThreshold={0.2}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={0.0001}
      stickyHeaderIndices={[]}
      viewabilityConfigCallbackPairs={[]}
    >
      <View>
        <View
          onFocusCapture={[Function]}
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              {
                "backgroundColor": "white",
                "height": 60,
                "margin": 20,
              }
            }
          >
            <Text>
              0
            </Text>
            <Text
              style={
                {
                  "fontSize": 8,
                }
              }
            >
              Art 1
            </Text>
            <Text>
              1
            </Text>
          </View>
        </View>
        <View
          onFocusCapture={[Function]}
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              {
                "backgroundColor": "white",
                "height": 60,
                "margin": 20,
              }
            }
          >
            <Text>
              1
            </Text>
            <Text
              style={
                {
                  "fontSize": 8,
                }
              }
            >
              Art 2
            </Text>
            <Text>
              2
            </Text>
          </View>
        </View>
        <View
          onLayout={[Function]}
        />
      </View>
    </RCTScrollView>
  </View>
</RCTSafeAreaView>
`;
